### PR TITLE
Add ability to restart and clear all.

### PR DESCRIPTION
**Pull request by Carreau**
It is a oft requested feature.

----

Ping @ellisonbg , @minrk, @wrobstory, @phobson

### CODE REVIEW

**Comment by takluyver** [comment]
We had also talked about, when a fresh kernel is started and you execute the first cell, maybe all output from previous execution should be cleared, to avoid confusion with the results from the current kernel. Thoughts? I can see some downsides, but that might be nicer UX than giving the user another choice.

File: notebook/static/notebook/js/notebook.js [hunk 1]
```diff
@@ -1631,11 +1631,18 @@ define(function (require) {
             ),
             buttons : {
                 "Continue running" : {},
-                "Restart" : {
+                "Clear all cells output & restart" : {
```
Comment (minrk) [inline notebook/static/notebook/js/notebook.js]: s/cells//

Comment (ellisonbg) [inline notebook/static/notebook/js/notebook.js]: Just finishing a final will look at this soon...

On Wed, Jun 10, 2015 at 3:23 PM, Min RK <notifications@github.com> wrote:

> In notebook/static/notebook/js/notebook.js
> <https://github.com/jupyter/notebook/pull/148#discussion_r32173261>:
>
> > @@ -1631,11 +1631,18 @@ define(function (require) {
> >              ),
> >              buttons : {
> >                  "Continue running" : {},
> > -                "Restart" : {
> > +                "Clear all cells output & restart" : {
>
> s/cells//
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/148/files#r32173261>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by Carreau** [comment]
> e had also talked about, when a fresh kernel is started and you execute the first cell, maybe all output from previous execution should be cleared, to avoid confusion with the results from the current kernel. Thoughts? I can see some downsides, but that might be nicer UX than giving the user another choice.

I don't want to change behavior like that so close to release. I don't like this behavior.

**Comment by ellisonbg** [comment]
I don't think it makes sense to change the default on this. However at one
point I did make the keyboard shortcut "000" do the clear output+restart
kernel action. I would also be fine with a button that does this action.


On Wed, Jun 10, 2015 at 3:26 PM, Matthias Bussonnier <
notifications@github.com> wrote:

> e had also talked about, when a fresh kernel is started and you execute
> the first cell, maybe all output from previous execution should be cleared,
> to avoid confusion with the results from the current kernel. Thoughts? I
> can see some downsides, but that might be nicer UX than giving the user
> another choice.
>
> I don't want to change behavior like that so close to release. I don't
> like this behavior.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/148#issuecomment-110935812>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by Carreau** [comment]
000 would shadow 00 which prompt for kernel restarting. A button is too much IMHO.

**Comment by jdavidheiser** [comment]
As a user, I don't like the idea of automatically clearing notebook output.  I've had some very long-running processes that generated graphs that could then be referenced through several iterations of restarting the kernel after other long-running computations had to be killed.  If I accidentally forgot that running a new cell would clear my output, and had to wait a few hours for a graph to be generated again, that would be a huge annoyance.  That's certainly an argument for me being smarter about saving my work, but automatically clearing output can cause a lot of frustration in general.

I do, however, like the idea of somehow obviously marking cells that were generated by "old" kernels, to reduce the confusion @takluyver mentioned.

The button / menu item approach seems ideal to me.

**Comment by Carreau** [comment]
We discussed a long time ago about "old" kernel and dirty cell. We arrived to the conclusion that there is no sigle, nor correct definition of what that would mean, neither a reliable way to detect it, unless we do some relatively large refactor of the notebook architecture(that we plan to do at some point anyway)

**Comment by Carreau** [comment]
Any thought on that ? It does not change default behavior, nor shortcut, just add a convenient button in the restart dialog, that also clear the cells...

**Comment by minrk** [comment]
I think this is fine.

**Comment by damianavila** [comment]
LGTM too...

**Comment by Carreau** [comment]
OK, let's make this live. I'll merge in the next timeI come across this if no complaints.

**Pull request merged**